### PR TITLE
feat: Implement desktopCapturer.setSkipCursor

### PR DESCRIPTION
**Pull request by adityak368**
#### Description of Change

feat: Implement electron.DesktopCapturer.setSkipCursor

Useful to dynamically stop cursor capture in the stream.
By default the cursor is captured.
First argument is a deviceId so one can stop capturing
the cursor for a specific streamed source.

electron#23923

This is essentially a cherrypick of the the previous implementation of @CapOM

https://github.com/electron/electron/pull/25103

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/master/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Implemented DesktopCapturer.setSkipCursor API. Useful to dynamically stop cursor capture in the stream.

### CODE REVIEW

**Comment by adityak368** [comment]
Note: I am not familiar with chromium and electron development. But would like to help as much as I can and get this over the line and merged

**Comment by nmilev-cursive** [comment]
@adityak368 If there's anything you need help with to make this PR successful, let me know!

**Comment by adityak368** [comment]
@nmilev-cursive @CapOM  I have just recently started looking into chromium development. I had to change the patch a bit so that it would hopefully pass the builds and work properly. Would be nice if you could have a look at the changes made (especially in the patch) and let me know if i missed something or could do something better

**Comment by ns-jisorce** [comment]
Try to run the tests locally on your machine, I believe you will be able to reproduce the crashes

File: lib/browser/rpc-server.ts
Comment (zcbenz) [inline lib/browser/rpc-server.ts:77]: Can you add documentation for this event?

**Review by zcbenz — COMMENTED** [review]
Can you add some tests? We don't have to test its functionality, it would be helpful to know it does not crash after Chromium upgrades.

File: lib/browser/rpc-server.ts
Comment (adityak368) [reply lib/browser/rpc-server.ts:77]: Sure. Added the documentation for the event. Added basic tests for the api.

**Comment by adityak368** [comment]
@ns-jisorce I only have a Windows machine. I ran the tests locally there. only 1 test failed (minimize restore window) which seems to be unrelated to the change. Also the tests failing in linux and mac builds seem to be unrelated to the change

**Comment by danielehrhardt** [comment]
Tested this on Mac. Looks good.

**Review by zcbenz — APPROVED** [review]

**Comment by zcbenz** [comment]
This PR requires review from @electron/wg-upgrades, and requires API review from @electron/wg-api .

**Comment by ns-jisorce** [comment]
The change looks good to me, if you can restore the author on the first commit that would be nice. (in principle it would have been preserved if you did git cherry-pick)
```
git commit --amend --no-edit --author="Julien Isorce <julien.isorce@gmail.com>"
```

Thanks!

**Review by nornagon — CHANGES_REQUESTED** [review]
This is a significant sized patch with no champion in the maintainers group, and no functional tests. It is thus very likely to cause issues when upgrading in future.

There are a few possible paths forward for this to get merged. The first and best is to see if there is a way to achieve the same result without requiring a patch. I am not personally familiar with the code, but I think this is worth investigating. 

If there really is no way to achieve this functionality without a patch, the next step would be to see if we can drastically reduce the size of the patch. Smaller patches are easier to maintain, because touching less code means the patch is less likely to conflict when upgrading.

If the patch size cannot be significantly reduced, the next thing to try would be to submit these patches upstream. See the [chromium contributing guidelines](https://chromium.googlesource.com/chromium/src/+/master/docs/contributing.md) and [WebRTC contributing guidelines](https://webrtc.org/support/contributing/) for how to get started with that.

If you try all the above and do not meet success, the final option is to significantly expand documentation and test coverage for these patches. The goal here would be that if a maintainer has to handle fixing this patch, then the tests should be able to report whether or not the feature is still working as intended, and the maintainer should be able to clearly understand the intention of each part of the patch, in order to be able to correctly resolve conflicts.

Thanks for your patience with this, I know it's a lot of work. Our maintainer team is small and we have to carefully consider the trade-offs of adding patches to the codebase. Every patch adds to the ongoing maintenance burden of keeping Electron up-to-date, and we have to ensure that the project remains sustainable.

**Comment by adityak368** [comment]
The topic of upstreaming this was already discussed before in the issue and we concluded that patching electron was better. Not sure if something has changed since then. The webrtc part of the patch is pretty straightforward in my opinion. The chromium part is maybe a bit more involved. Other changes are basically for the electron API which is also pretty straightforward. 

The docs for the API and its functionality are already present. Basic tests are written to know that it does not crash. Please let me know what other documentation you have in mind. Also I am not sure how we can write tests to check the functionality as the cursor is pretty much embedded in the video stream. Any suggestions and help to achieve that is welcome. 

@ns-jisorce What are your thoughts? @nornagon Please suggest how to move forward

**Comment by nornagon** [comment]
@adityak368 surely, patching Electron is easier for _you_, but it is not easier for the maintainers. For context, we currently have 110 patches against Chromium, and if merged this patch would be the 6th-largest patch. Chromium constantly changes, and we regularly have to update our patches to keep pace. The more and larger patches we have, the larger our ongoing maintenance burden.

I've outlined several possible paths for moving forward already. The PR adds events which are never tested, nor is calling the method while a capture is occurring. Calling the method from the renderer is not tested either. The documentation this PR adds is pretty thin. There is no example code, it's not documented whether the events are emitted when calling `setSkipCursor` from the main process or only from the renderer, and there is only the barest explanation of what the feature does. We don't need an essay of docs for the feature, as it's fairly straightforward, but it could be better. If a maintainer with no context about what this feature is intended to do is expected to fix a 350-line patch in order to keep the feature working, they had better be able to understand what "working" means.

File: lib/browser/desktop-capturer.ts
Comment (nornagon) [inline lib/browser/desktop-capturer.ts:85]: This is a memory leak. `createDesktopCapturer()` returns an object which is not released until capturing has finished.

File: shell/browser/api/electron_api_desktop_capturer.cc
Comment (nornagon) [inline shell/browser/api/electron_api_desktop_capturer.cc:130]: since this is a static method, which does not need a `DesktopCapturer` object for context, it should be exposed directly on the bindings object in `Initialize`.

**Review by lastthyme — APPROVED** [review]

**Review by lastthyme — APPROVED** [review]

File: shell/browser/api/electron_api_desktop_capturer.cc
Comment (adityak368) [reply shell/browser/api/electron_api_desktop_capturer.cc]: Done. Removed the method from DesktopCapturer and moved it to Initialize

File: lib/browser/desktop-capturer.ts
Comment (adityak368) [reply lib/browser/desktop-capturer.ts]: Not creating a desktopcapturer anymore. calling the static method directly

**Comment by adityak368** [comment]
@nornagon Addressed review comments. Removed the events. Added more documentation. Added sample code. Added test to call setskipcursor during the capture.

**Comment by adityak368** [comment]
@nornagon Please let me know if there everything is fine or if there are any other changes required.

**Comment by hackers365** [comment]
up

**Review by danielehrhardt — APPROVED** [review]

**Comment by michalzaq12** [comment]
up

**Comment by danielehrhardt** [comment]
Why did you closed the Pull Request?

**Pull request closed without merging**
